kernel/src/mem.rs

```diff
@@ -1,202 +1,278 @@
+    /// ```
+    /// use kernel::ReadWriteAppSlice;
+    /// const DEFAULT_RWAPPSLICE_VAL: ReadWriteAppSlice = ReadWriteAppSlice::const_default();
+    /// let my_array = [DEFAULT_RWAPPSLICE_VAL; 12];
+    /// ```
+    pub const fn const_default() -> Self {
+        Self {
+            ptr: 0x0 as *mut u8,
+            len: 0,
+            process_id: None,
+        }
+    }
+}
+
+impl Default for ReadWriteAppSlice {
+    fn default() -> Self {
+        Self::const_default()
+    }
+}
+
+impl ReadWrite for ReadWriteAppSlice {
+    fn mut_ptr(&self) -> *mut u8 {
+        if self.len == 0 {
+            0x0 as *mut u8
+        } else {
             self.ptr
-                .process
-                .kernel
-                .process_map_or(false, appid, |process| {
-                    process
-                        .add_mpu_region(self.ptr() as *const u8, self.len(), self.len())
-                        .is_some()
-                })
+        }
+    }
+
+    fn mut_map_or<F, R>(&self, default: R, fun: F) -> R
+    where
+        F: FnOnce(&mut [u8]) -> R,
+    {
+        match self.process_id {
+            None => default,
+            Some(pid) => pid.kernel.process_map_or(default, pid, |_| {
+                // Safety: `kernel.process_map_or()` validates that the process still exists
+                // and its memory is still valid. `Process` tracks the "high water mark" of
+                // memory that the process has `allow`ed to the kernel, and will not permit
+                // the process to free any memory after it has been `allow`ed. This guarantees
+                // that the buffer is safe to convert into a slice here.
+                let slice = unsafe { slice::from_raw_parts_mut(self.ptr, self.len) };
+                fun(slice)
+            }),
+        }
+    }
+}
+
+impl Read for ReadWriteAppSlice {
+    fn len(&self) -> usize {
+        self.process_id
+            .map_or(0, |pid| pid.kernel.process_map_or(0, pid, |_| self.len))
+    }
+
+    fn ptr(&self) -> *const u8 {
+        if self.len == 0 {
+            0x0 as *const u8
         } else {
-            false
+            self.ptr
         }
     }
 
-    /// Returns an iterator over the slice
-    ///
-    /// See
-    /// [`std::slice::iter()`](https://doc.rust-lang.org/std/primitive.slice.html#method.iter).
-    ///
-    /// Internally this uses
-    /// [`AsRef`](struct.AppSlice.html#impl-AsRef<[T]>), hence when
-    /// the app dies, restarts or the
-    /// [`AppId`](crate::callback::AppId) changes for any other
-    /// reason, the iterator will be of zero length.
-    pub fn iter(&self) -> slice::Iter<T> {
-        self.as_ref().iter()
+    fn map_or<F, R>(&self, default: R, fun: F) -> R
+    where
+        F: FnOnce(&[u8]) -> R,
+    {
+        match self.process_id {
+            None => default,
+            Some(pid) => pid.kernel.process_map_or(default, pid, |_| {
+                let slice = unsafe { slice::from_raw_parts(self.ptr, self.len) };
```

Comment:
`unsafe` comment? likely the same as the `mut_map_or` use above?

Reply:
This should be fixed in #2464.